torchacc/utils/patch.py

```diff
@@ -92,3 +93,38 @@ def update_causal_mask(
             return None
 
         LlamaModel._update_causal_mask = update_causal_mask
+
+
+def patch_qwen():
+    '''
+    Modify the calculation of `rotary_seq_len` in `Qwen2FlashAttention2.forward` to avoid xla graph be executed
+    '''
+    import inspect
+    import transformers
+    from .logger import logger
+    from packaging import version
+    if version.parse(transformers.__version__) >= version.parse("4.37.0"):
+        try:
+            import transformers.models.qwen2.modeling_qwen2 as qwen2
+            import re
+
+            src = inspect.getsource(qwen2.Qwen2FlashAttention2)
+
+            pattern1 = r"rotary_seq_len\s*=\s*\(\s*max\(kv_seq_len,\s*position_ids\[:,\s*-1\]\.max\(\)\.item\(\)\s*\+\s*1\)\s*if\s*position_ids\s*is\s*not\s*None\s*else\s*kv_seq_len\s*\)"
+            pattern2 = r"rotary_seq_len\s*=\s*max\(kv_seq_len,\s*position_ids\[:,\s*-1\]\.max\(\)\.item\(\)\)\s*\+\s*1"
+            replacement = "rotary_seq_len = kv_seq_len"
+
+            src = re.sub(pattern1, replacement, src)
+            src = re.sub(pattern2, replacement, src)
+            dict_src = \
+            """
+    QWEN2_ATTENTION_CLASSES = {
+        "eager": Qwen2Attention,
+        "flash_attention_2": Qwen2FlashAttention2,
+        "sdpa": Qwen2SdpaAttention,
+    }
+            """
+            src = src + dict_src
+            exec(src, qwen2.__dict__)
+        except Exception as e:
+            logger.warning(f"patch qwen2 failed due to: {e}")
```

Comment:
better to add log when transformers.__version__ < 4.37.0

Reply:
Since qwen2 model is added to transformers in version 4.37.0，I think it's fine not print anything when version < 4.37.0

Reply:
ok